add log_param/log_params

dtrifiro opened the pull request:
Provides two new methods which log parameters to a yaml file (`params.yaml`):

- `log_param(name, value)` logs the given parameter name/value
- `log_params(params)` logs the given set of parameters (dict).

shcheklein commented:
Should we log into `params.yaml` ? To simplify the migration and so that we can show them as params in Studio/VS Code?

Otherwise, what is the plan to support this in both - Studio/VS Code?

daavoo commented on src/dvclive/live.py:
```diff
@@ -79,6 +85,7 @@ def _cleanup(self):
             shutil.rmtree(
                 Path(self.dir) / data_type.subfolder, ignore_errors=True
             )
+        shutil.rmtree(Path(self.dir) / self.PARAMS_DIR, ignore_errors=True)
```
Why not handle whether or not to remove the dir base on `self._resume` here instead of inside the `log_param` method?

dtrifiro replied:
Not sure what you mean here

daavoo commented on src/dvclive/live.py:
```diff
@@ -191,6 +198,78 @@ def log_plot(self, name, labels, predictions, **kwargs):
+            logger.debug("resume=True, not removing existing parameters.")
+            return
+
+        output_dir = Path(self.dir) / self.PARAMS_DIR
+        output_dir.mkdir(exist_ok=True)
+        params_path = output_dir / (
+            f"{name}.json" if not name.endswith(".json") else name
+        )
```
(Implementation)
Not a blocker but, for consistency, I think would be nice if we implement a new `Data` subclass (i.e. `Param`) and move this logic there.

It would look more similar to the other `log_x` methods and could launch explicit exceptions for invalid data types, etc.

dtrifiro replied:
I thought about that, and that's the first approach I went for, although if you look at the `Data` class most of its (abstract) methods are tightly connected to the "stepped" workflow:  `step`, `no_step_output_path`, `first_step_dump`, `no_step_dump`, `step_dump`. Maybe have a new superclass from which `Data` and `Param` derive, although I think they would share only few methods/attributes.

I think a `Param` class is something we will have to add, especially in relation to the nested dictionaries/complex objects parameters support.

daavoo commented on src/dvclive/live.py:
```diff
@@ -191,6 +198,78 @@ def log_plot(self, name, labels, predictions, **kwargs):
         data.dump(val, self._step, **kwargs)
         logger.debug(f"Logged {name}")
 
+    def log_params(
+        self, params: Dict[str, ParamsLike], name: str = PARAMS_DEFAULT_NAME
+    ):
```
(Product) I am not sure about including this one (at least until we also include equivalents for other data types). 

(Implementation) Regardless, is there a reason for this method to don't reuse `log_param` internally?

dberenbaum replied:
I think it's okay to start here because we have to start somewhere and params are maybe the most likely scenario for having to log many values at once.

However, can we collapse into one method that accepts either? It's messier to implement but less for users to remember.

dtrifiro replied:
> (Implementation) Is there a reason for this method to don't reuse log_param internally?

These were slighty different initially, but the ended up converging, I'll change `log_param` so that it uses `log_params` internally.

dtrifiro replied:
> However, can we collapse into one method that accepts either? It's messier to implement but less for users to remember.

Wouldn't it be confusing? As in: with a single method, if logging a `dict`, are you logging a single parameter that takes a `dict` value (this is somehow related to your nested params comment) or are you logging a group of parameters? Maybe it would be just enough to provide an optional `name` argument to distinguish between the two cases, although I stil feel naming would be odd.

Something like this, perhaps: 
```python
dvclive.log_param(name="optimizer_kwargs", {"learning_rate": 1e-3, "name": "adam"}) # this logs a dict parameter
dvclive.log_param({"learning_rate": 1e-3, "optimizer_name": "adam"}) # this logs a dict parameter
```

dtrifiro replied:
Any thoughts here?

dberenbaum replied:
LGTM with `log_param` and `log_params`, thanks!

dberenbaum replied:
Still seems a little odd because of how similar they both are now, but it's probably the most familiar approach for both dvclive and mlflow users, and it gives the most flexibility to users, so I'm good with it.

daavoo commented on src/dvclive/live.py:
```diff
@@ -191,6 +198,78 @@ def log_plot(self, name, labels, predictions, **kwargs):
+        )
+
+        if params_path.is_file():
+            with open(params_path, "r", encoding="utf-8") as fobj:
+                loaded_params = json.load(fobj)
```
(Product) Should we load the existing params? I think we could just treat as the other things we log and overwrite unless `resume` is used

dtrifiro replied:
This depends on how we want to handle `log_param`/`log_params`. If using `log_param(name, val)` I would expect this to append the parameter to an existing file. Using `log_params` could avoid loading the parameters, although it could be confusing to have different behaviors. 

Note that the params directory is always cleared on dvclive's init and we are not dumping parameters when `resume==True`

dtrifiro replied:
Changed this slightly, now skipping `log_param/log_params` when `resume==True an step > 0`

daavoo commented on src/dvclive/live.py:
```diff
@@ -191,6 +198,78 @@ def log_plot(self, name, labels, predictions, **kwargs):
+        self,
+        name: str,
+        val: Union[int, float, str, bool],
+        params_set_name: str = PARAMS_DEFAULT_NAME,
```
(Product) I am not sure about providing this option. 

I believe we could start by "forcing" a fixed output file of our choice so that the Onboarding workflow gets easier to implement.

We can wait for users to ask for this unless we see a good reason to include the logic as part of the current onboarding proposal (right now feels a little like a deviation from what we have been discussing)

dtrifiro replied:
That's why this is optional ("forcing" a fixed output). The idea is to let users log parameters to different files to keep a flatter structure in each file. Could get rid of this though.

dberenbaum replied:
Seems like it deviates from the rest of dvclive where the output structure is fixed relative to the dvclive path, and it might be confusing to document, so my preference is to leave this out until we have a need.

dtrifiro replied:
Got rid of it

daavoo commented on src/dvclive/live.py:
```diff
@@ -22,9 +23,13 @@
 logger = logging.getLogger("dvclive")
 logger.setLevel(os.getenv(env.DVCLIVE_LOGLEVEL, "INFO").upper())
 
+ParamsLike = Union[int, float, str, bool]
```
❤️ . Would be awesome to also add this to the other methods `log`, `log_plot` `log_image` (not as part of this P.R)

dberenbaum commented on src/dvclive/live.py:
```diff
@@ -191,6 +198,78 @@ def log_plot(self, name, labels, predictions, **kwargs):
         logger.debug(f"Logged {name}")
 
+    def log_params(
+        self, params: Dict[str, ParamsLike], name: str = PARAMS_DEFAULT_NAME
```
What about handling nested dicts? Similarly, can `log_param("train.lr", 0.1)` be saved as a nested dict?

dtrifiro replied:
We should be handling those, although I think this can become tricky: think of more complex objects, such as an optimizer (like in here https://keras.io/api/optimizers/). How are we supposed to capture these kind of parameters?

daavoo replied:
I would expect people to log that kind of thing as follows:

`Live().log_params("train", {"lr": 0.01})`

Which works, although I don't know if unexpectedly 😅 as it doesn't match the `ParamsLike` type @dtrifiro

dtrifiro replied:
Yep, `ParamsLike` should really look like this: `ParamsLike = Union[int, float, str, bool, "ParamsLike"]`, although mypy complains because of a cyclic definition. Will check this out later.

dberenbaum replied:
Sorry, for some reason I thought the dot-notation was supported for logging metrics. That shouldn't be necessary, but handling nested dicts should be.

Some interesting reading:
* https://stackoverflow.com/questions/68167770/type-hints-for-variably-nested-dict-in-python
* https://github.com/python/mypy/issues/731

dtrifiro replied:
Fixed `ParamsLike` to include the recursive definition. For now we support dumping any parameters that are supported by `json.dump`. Complex (custom) parameters are still not supported. Do we want to support those right away, or can we leave those to another iteration?

dberenbaum commented:
> Should we log into `params.yaml` ? To simplify the migration and so that we can show them as params in Studio/VS Code?

I'm not sure if it simplifies migration to put them into `params.yaml`, since it then requires understanding that you need to remove your `log_param()` calls or end up overwriting your param dependencies. Saving them in some other YAML file might be less confusing, although it looks a bit odd to be saving data in both YAML and JSON (no strong opinion there). 

> Otherwise, what is the plan to support this in both - Studio/VS Code?

This is still under discussion, but a custom params file could be supported through `dvc.yaml` config or through some DVCLive convention.

dtrifiro commented:
> I'm not sure if it simplifies migration to put them into params.yaml, since it then requires understanding that you need to remove your log_param() calls or end up overwriting your param dependencies.

+1 

> Saving them in some other YAML file might be less confusing, although it looks a bit odd to be saving data in both YAML and JSON (no strong opinion there).

One reason I see for preferring YAML is to provide users with an example on how a "standard" `params.yaml` file would look like, although this does not provide a clear path on how to go forward from there: "ok, I have a `params.yaml` file which I could use as a params dependency, although I still have to rewrite my code in order to read/use that file"

dtrifiro commented:
Note: linting failures on mac/win 3.10 seem unrelated

dberenbaum commented:
> Should we log into `params.yaml` ?

@dtrifiro Do you want to migrate to a YAML file? I actually don't think `params.yaml` is that problematic as long as it's inside the dvclive dir.

dtrifiro commented on src/dvclive/live.py:
```diff
@@ -23,8 +24,16 @@
+
 class Live:
     DEFAULT_DIR = "dvclive"
+    PARAMS_DIR = "params"
```
Not if we do not want multiple parameters files, will remove

dberenbaum commented:
Discussed with @dtrifiro and decided that we should be able to implement basic YAML parsing without copying all of the DVC YAML serialization utilities.

daavoo commented on src/dvclive/live.py:
```diff
@@ -79,6 +93,8 @@ def _cleanup(self):
             shutil.rmtree(
                 Path(self.dir) / data_type.subfolder, ignore_errors=True
             )
+        if os.path.exists(self.params_path):
+            os.remove(self.params_path)
 
         for f in {self.summary_path, self.report_path}:
             if os.path.exists(f):
```
```suggestion
        for f in {self.summary_path, self.report_path, self.params_path}:
            if os.path.exists(f):
```

daavoo commented on src/dvclive/live.py:
```diff
@@ -191,6 +211,43 @@ def log_plot(self, name, labels, predictions, **kwargs):
         logger.debug(f"Logged {name}")
 
+    def log_params(self, params: Dict[str, ParamLike]):
+        """Saves the given set of parameters (dict) to json"""
```
```suggestion
        """Saves the given set of parameters (dict) to `self.params_path`"""
```

daavoo commented on src/dvclive/live.py:
```diff
@@ -191,6 +211,43 @@ def log_plot(self, name, labels, predictions, **kwargs):
 
+    def log_params(self, params: Dict[str, ParamLike]):
+        """Saves the given set of parameters (dict) to json"""
+        if self._resume and self._step:
+            logger.info(
+                "Resuming previous dvclive session, not logging params."
+            )
+            return
```
Missing test for this

daavoo commented on src/dvclive/serialize.py:
```diff
@@ -0,0 +1,47 @@
+    return yaml
+
+
+def _dump(data, stream):
+    yaml = _get_yaml()
+    return yaml.dump(data, stream)
```
Unused?
```suggestion
```

daavoo reviewed (changes requested):
Some minor comments and good to go.

Main doubt:

(Product) I am missing the point of `YAML` vs `JSON` but I have no strong opinion. I just think the right direction would be to implement https://github.com/iterative/dvc/issues/7939 and save us the new dependency / extra utils code.

Must address:

(Implementation) @dtrifiro Need to either make all imports of `ruamel.yaml` optional with `try/except ImporError` or add as a `install_requires` dependency.

dberenbaum commented:
> (Product) I am missing the point of `YAML` vs `JSON` but I have no strong opinion. I just think the right direction would be to implement [iterative/dvc#7939](https://github.com/iterative/dvc/issues/7939) and save us the new dependency / extra utils code.

@daavoo Not sure I understand your point (I also have no strong opinion on `YAML` vs `JSON`). How does a configurable params file save code/dependencies here?

daavoo commented:
> @daavoo Not sure I understand your point (I also have no strong opinion on `YAML` vs `JSON`). How does a configurable params file save code/dependencies here?

If I understood the motivation for `params.yaml` was to `simplify the migration`, right? 
If we allow DVC to set custom default params file, migration would be equally simple with `json` format

dtrifiro commented:
Addressed the remaining comments. 

> If we allow DVC to set custom default params file
Do we really need this? We can already specify custom params files, although we "nudge" users towards using `params.yaml`

dberenbaum commented:
> If I understood the motivation for `params.yaml` was to `simplify the migration`, right?
> If we allow DVC to set custom default params file, migration would be equally simple with `json` format

Good point. I would say YAML is arguably more easily readable and editable and maybe more common for parameters, but JSON is not so different, has fewer issues, and is already used for metrics, so 🤷 . Let's merge this and we can come back to it if needed in the future.
